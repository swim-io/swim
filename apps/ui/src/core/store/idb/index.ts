import type { Table } from "dexie";
import Dexie from "dexie";

import type { Env } from "../../../config";
import type { InteractionState } from "../../../models";

import type { PersistedInteractionState } from "./helpers";
import {
  deserializeInteractionStates,
  prepareInteractionState,
} from "./helpers";

const MAX_STORED_INTERACTIONS = 10;

type IDBState = {
  readonly id: string;
} & PersistedInteractionState;
export class SwimIDB extends Dexie {
  interactionStates!: Table<IDBState, string>;

  constructor() {
    super("SwimIDB");
    this.version(1).stores({ interactionStates: "&id" });
  }

  async getInteractionStates(env: Env) {
    return this.transaction("rw", "interactionStates", async () => {
      const data = await this.interactionStates
        .filter((idbState) => idbState.interaction.env === env)
        .toArray();
      return this.deserializeState(data, env);
    }).catch((err) => {
      console.warn(err);
    });
  }

  addInteractionState(interactionState: InteractionState) {
    this.transaction("rw", "interactionStates", async () => {
      const serializedState: PersistedInteractionState =
        this.serializeState(interactionState);
      await this.interactionStates.add(
        {
          id: interactionState.interaction.id,
          ...serializedState,
        },
        interactionState.interaction.id,
      );
      const size = await this.interactionStates.count();
      if (size > MAX_STORED_INTERACTIONS) {
        const diff = size - MAX_STORED_INTERACTIONS;
        const collection = await this.interactionStates.toArray();
        for (let index = 0; index < diff; index++) {
          await this.interactionStates.delete(collection[index].id);
        }
      }
    }).catch((err) => {
      console.warn("Fail to add interactionState into idb", err);
    });
  }

  private serializeState(state: InteractionState): PersistedInteractionState {
    return prepareInteractionState(state);
  }

  private deserializeState(
    state: readonly PersistedInteractionState[],
    env: Env,
  ): readonly InteractionState[] {
    return deserializeInteractionStates(state, env);
  }
}

export const idb = new SwimIDB();

export function resetDatabase() {
  return idb.transaction("rw", idb.interactionStates, async () => {
    await Promise.all(idb.tables.map((table) => table.clear()));
  });
}
