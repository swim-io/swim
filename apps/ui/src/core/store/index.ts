export * from "./useNotification";
export * from "./useEnvironment";
export * from "./useWalletAdapter";
export * from "./useInteractionState";
export * from "./useInteractionStateV2";
