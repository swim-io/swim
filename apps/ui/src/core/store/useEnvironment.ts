import { produce } from "immer";
import create from "zustand";
import type { GetState, SetState, StoreApi } from "zustand";
import type { StateStorage } from "zustand/middleware";
import { persist } from "zustand/middleware.js";

import type { Env } from "../../config";
import { DEFAULT_ENV, isValidEnv } from "../../config";

export interface EnvironmentState {
  readonly env: Env;
  readonly customLocalnetIp: string | null;
  readonly setEnv: (newEnv: Env) => void;
  readonly setCustomLocalnetIp: (ip: string | null) => void;
}

export const useEnvironment = create(
  persist<EnvironmentState>(
    (
      set: SetState<EnvironmentState>,
      get: GetState<EnvironmentState>,
      api: StoreApi<EnvironmentState>,
    ) => ({
      env: DEFAULT_ENV,
      customLocalnetIp: null,
      setEnv: (newEnv: Env) => {
        set(
          produce<EnvironmentState>((draft) => {
            if (api.getState().customLocalnetIp !== null) {
              draft.env = newEnv;
            }
          }),
        );
      },
      setCustomLocalnetIp: (ip: string | null) => {
        set(
          produce<EnvironmentState>((draft) => {
            draft.customLocalnetIp = ip;
            draft.env = isValidEnv(api.getState().env)
              ? api.getState().env
              : DEFAULT_ENV;
          }),
        );
      },
    }),
    {
      name: "env-config",
      getStorage: (): StateStorage => localStorage,
      partialize: (state: EnvironmentState) => ({
        env: state.env,
        customLocalnetIp: state.customLocalnetIp,
      }),
      merge: (
        persistedState: any, // TODO: Set type to unknown and validate
        currentState: EnvironmentState,
      ): EnvironmentState => {
        const { env, customLocalnetIp } = persistedState;
        if (isValidEnv(env)) {
          if (customLocalnetIp !== null) {
            return { ...currentState, env, customLocalnetIp };
          }
          return { ...currentState, ...persistedState };
        }
        return currentState;
      },
    },
  ),
);
