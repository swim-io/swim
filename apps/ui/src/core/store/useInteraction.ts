/* eslint-disable functional/immutable-data */
import { createDraft, produce } from "immer";
import type { Draft } from "immer";
import type { GetState, SetState, StoreApi } from "zustand";
import create from "zustand";
import type { StateStorage } from "zustand/middleware";
import { persist } from "zustand/middleware.js";

import type {
  FromSolanaTransferState,
  Interaction,
  RequiredSplTokenAccounts,
  SolanaPoolOperationState,
  ToSolanaTransferState,
} from "../../models";

export interface InteractionState {
  readonly interactions: readonly Interaction[];
  readonly requiredSplTokenAccounts: RequiredSplTokenAccounts;
  readonly toSolanaTransfers: readonly ToSolanaTransferState[];
  readonly solanaPoolOperations: readonly SolanaPoolOperationState[];
  readonly fromSolanaTransfers: readonly FromSolanaTransferState[];
  readonly currentInteractionIndex: number;
  readonly addInteraction: (interaction: Interaction) => void;
  readonly setCurrentInteraction: (interactionId: string) => void;
}

export const createInteractionSlice = create(
  persist<InteractionState>(
    (
      set: SetState<InteractionState>,
      get: GetState<InteractionState>,
      api: StoreApi<InteractionState>,
    ) => ({
      interactions: [],
      requiredSplTokenAccounts: {},
      toSolanaTransfers: [],
      solanaPoolOperations: [],
      fromSolanaTransfers: [],
      currentInteractionIndex: 0,
      addInteraction: (interaction: Interaction) => {
        set(
          produce<InteractionState>((draft) => {
            draft.interactions.push(createDraft(interaction));
          }),
        );
      },
      setCurrentInteraction: (interactionId) => {
        set(
          produce<InteractionState>((draft) => {
            draft.currentInteractionIndex = draft.interactions.findIndex(
              (interaction) => interaction.id === interactionId,
            );
          }),
        );
      },
    }),
    {
      name: "interaction-db",
      getStorage: (): StateStorage => localStorage,
    },
  ),
);
