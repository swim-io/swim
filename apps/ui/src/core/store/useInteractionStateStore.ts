/* eslint-disable functional/immutable-data */
import { createDraft, produce } from "immer";
import type { WritableDraft } from "immer/dist/internal";
import type { SetState } from "zustand";
import create from "zustand";

import {
  EcosystemId,
  getSolanaTokenDetails,
  isEvmEcosystemId,
} from "../../config";
import { useConfig, useEnvironment, useSolanaWallet } from "../../contexts";
import {
  usePoolMaths,
  useSplTokenAccountsQuery,
  useWallets,
} from "../../hooks";
import {
  useRequiredPoolsForInteraction,
  useRequiredTokensForInteraction,
} from "../../hooks/interaction";
import type {
  FromSolanaTransferState,
  Interaction,
  InteractionSpec,
  InteractionState,
  RequiredSplTokenAccounts,
  SolanaPoolOperationState,
  ToSolanaTransferState,
} from "../../models";
import {
  InteractionType,
  createOperationSpecs,
  findTokenAccountForMint,
  generateId,
  getConnectedWallets,
  getRequiredPools,
  getTokensByPool,
} from "../../models";
import { isNotNull } from "../../utils";

export interface InteractionStateStore {
  readonly interactionStateStore: readonly InteractionState[];
  readonly createInteraction: (interactionSpec: InteractionSpec) => void;
  readonly updateInteractionState: (
    interactionId: string,
    updateFn: (draft: WritableDraft<InteractionState>) => void,
  ) => void;
}

const useCreateInteractionFromSpec = (
  interactionSpec: InteractionSpec,
): Interaction => {
  const interactionId = generateId();
  const config = useConfig();
  const wallets = useWallets();
  const { env } = useEnvironment();
  const requiredPools = getRequiredPools(config.pools, interactionSpec);
  const inputPool = requiredPools.find(Boolean) ?? null;
  if (inputPool === null) {
    throw new Error("Unknown input pool");
  }
  const connectedWallets = getConnectedWallets(
    config.tokens,
    interactionSpec,
    wallets,
  );
  return {
    ...interactionSpec,
    id: interactionId,
    poolIds: requiredPools.map((pool) => pool.id),
    env,
    submittedAt: Date.now(),
    signatureSetKeypairs: {},
    previousSignatureSetAddresses: {},
    connectedWallets,
  };
};

export const useRequiredSplTokenAccounts = (
  interaction: Interaction,
): RequiredSplTokenAccounts => {
  const requiredTokens = useRequiredTokensForInteraction(interaction);
  const mints = requiredTokens.map(
    (token) => getSolanaTokenDetails(token).address,
  );
  const { data: tokenAccounts = [] } = useSplTokenAccountsQuery();
  const { address: walletAddress } = useSolanaWallet();
  if (walletAddress === null) {
    throw new Error("No Solana wallet address found");
  }
  return mints.reduce((state, mint) => {
    const accountForMint = findTokenAccountForMint(
      mint,
      walletAddress,
      tokenAccounts,
    );
    return {
      ...state,
      [mint]: {
        isExistingAccount: accountForMint !== null,
        account: accountForMint,
        txId: null,
      },
    };
  }, {});
};

export const useCreateToSolanaTransfers = (
  interaction: Interaction,
): readonly ToSolanaTransferState[] => {
  // Only for swap at the moment
  if (interaction.type !== InteractionType.Swap) {
    return [];
  }
  const {
    params: { exactInputAmount },
  } = interaction;
  const fromToken = exactInputAmount.tokenSpec;
  const fromEcosystem = fromToken.nativeEcosystem;
  // No ToSolanaTransfer if fromToken is not Evm
  if (!isEvmEcosystemId(fromEcosystem)) {
    return [];
  }
  return [
    {
      token: fromToken,
      value: exactInputAmount.toHuman(fromEcosystem),
      txIds: {
        approveAndTransferEvmToken: null,
        postVaaOnSolana: null,
        claimTokenOnSolana: null,
      },
    },
  ];
};

export const useCreateSolanaPoolOperations = (
  interaction: Interaction,
): readonly SolanaPoolOperationState[] => {
  const config = useConfig();
  const tokensByPoolId = getTokensByPool(config);
  const pools = useRequiredPoolsForInteraction(interaction);
  const poolMaths = usePoolMaths(pools.map(({ id }) => id)).filter(isNotNull);
  const operationSpecs = createOperationSpecs(
    tokensByPoolId,
    pools,
    poolMaths,
    interaction,
  );
  return operationSpecs.map((operation) => ({
    operation,
    txId: null,
  }));
};

export const useCreateFromSolanaTransfers = (
  interaction: Interaction,
): readonly FromSolanaTransferState[] => {
  // Only for swap at the moment
  if (interaction.type !== InteractionType.Swap) {
    return [];
  }
  const {
    params: { minimumOutputAmount },
  } = interaction;
  const toToken = minimumOutputAmount.tokenSpec;
  // No WormholeFromSolana if toToken is Solana
  if (toToken.nativeEcosystem === EcosystemId.Solana) {
    return [];
  }
  return [
    {
      token: toToken,
      value: null,
      txIds: {
        transferSplToken: null,
        claimTokenOnEvm: null,
      },
    },
  ];
};

const useCreateInteractionState = (
  interactionSpec: InteractionSpec,
): InteractionState => {
  const interaction = useCreateInteractionFromSpec(interactionSpec);
  return {
    interaction,
    requiredSplTokenAccounts: useRequiredSplTokenAccounts(interaction),
    toSolanaTransfers: useCreateToSolanaTransfers(interaction),
    solanaPoolOperations: useCreateSolanaPoolOperations(interaction),
    fromSolanaTransfers: useCreateFromSolanaTransfers(interaction),
  };
};

export const useInteractionStateStore = create<InteractionStateStore>(
  (set: SetState<InteractionStateStore>) => ({
    interactionStateStore: [],
    createInteraction: (interactionSpec: InteractionSpec) => {
      set(
        produce<InteractionStateStore>((draft) => {
          const interactionState = useCreateInteractionState(interactionSpec);
          draft.interactionStateStore.push(createDraft(interactionState));
        }),
      );
    },
    updateInteractionState: (interactionId, updateFn) => {
      set(
        produce<InteractionStateStore>((draft) => {
          const index = draft.interactionStateStore.findIndex(
            ({ interaction }) => interaction.id === interactionId,
          );
          updateFn(draft.interactionStateStore[index]);
        }),
      );
    },
  }),
);
