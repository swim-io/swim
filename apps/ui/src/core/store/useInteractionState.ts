/* eslint-disable functional/immutable-data */
import type { Draft } from "immer";
import { castDraft, produce } from "immer";
import type { GetState, SetState } from "zustand";
import create from "zustand";

import type { Interaction, InteractionState } from "../../models";
import type { ReadonlyRecord } from "../../utils";

import { idb } from "./idb";

export interface InteractionStore {
  readonly errorMap: ReadonlyRecord<Interaction["id"], Error | undefined>;
  readonly interactionStates: readonly InteractionState[];
  readonly recentInteractionId: string | null;
  readonly getInteractionError: (id: string) => Error | undefined;
  readonly setInteractionError: (id: string, error: Error | undefined) => void;
  readonly getInteractionState: (interactionId: string) => InteractionState;
  readonly addInteractionState: (interactionState: InteractionState) => void;
  readonly getInteractionStatesFromIDB: () => Promise<
    void | readonly InteractionState[]
  >;
  readonly updateInteractionState: (
    interactionId: string,
    updateCallback: (interactionState: Draft<InteractionState>) => void,
  ) => void;
}

export const useInteractionState = create(
  (set: SetState<InteractionStore>, get: GetState<InteractionStore>) => ({
    errorMap: {},
    interactionStates: [],
    recentInteractionId: null,
    getInteractionError: (id: string) => get().errorMap[id],
    setInteractionError: (id: string, error: Error | undefined) => {
      set(
        produce<InteractionStore>((draft) => {
          draft.errorMap[id] = error;
        }),
      );
    },
    getInteractionState: (id: string) => {
      const interactionState = get().interactionStates.find(
        ({ interaction }) => interaction.id === id,
      );
      if (!interactionState) {
        throw new Error(`Interaction ${id} not exist`);
      }
      return interactionState;
    },
    getInteractionStatesFromIDB: async () => {
      const data = (await idb.getInteractionStates()) || [];
      set(
        produce<InteractionStore>((draft) => {
          draft.interactionStates = castDraft(data);
        }),
      );
    },
    addInteractionState: (interactionState) => {
      set(
        produce<InteractionStore>((draft) => {
          draft.interactionStates.push(castDraft(interactionState));
          draft.recentInteractionId = interactionState.interaction.id;
        }),
      );
      idb.addInteractionState(interactionState);
    },
    updateInteractionState: (interactionId, updateCallback) => {
      set(
        produce<InteractionStore>((draft) => {
          const index = draft.interactionStates.findIndex(
            ({ interaction }) => interaction.id === interactionId,
          );
          updateCallback(draft.interactionStates[index]);
        }),
      );
    },
  }),
);
