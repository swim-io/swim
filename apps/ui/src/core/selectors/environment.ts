import { Env } from "@swim-io/core";

import { CONFIGS } from "../../config";
import type { EnvironmentState } from "../store";

export const DEFAULT_ENV = Env.Mainnet;

export const selectEnvs = (state: EnvironmentState) =>
  state.customIp === null ? [DEFAULT_ENV] : Object.values(Env);

export const selectConfig = (state: EnvironmentState) => CONFIGS[state.env];
