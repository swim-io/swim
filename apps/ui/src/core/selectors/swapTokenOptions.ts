import type { TokenOption } from "../../models";
import { findOrThrow } from "../../utils";
import type { EnvironmentState } from "../store";

import { selectConfig } from "./environment";

const swimUsdRegExp = /-swimusd$/;
const isSwimUsdTokenId = (tokenId: string): boolean =>
  swimUsdRegExp.test(tokenId);

export const selectSwapTokenOptions = (
  state: EnvironmentState,
): readonly TokenOption[] => {
  const { pools, tokens } = selectConfig(state);
  const restructuredPools = pools.filter((pool) => !pool.isLegacyPool);
  const nonLpTokenSpecs = restructuredPools
    .filter((pool) => !pool.isStakingPool)
    .flatMap((pool) => pool.tokens)
    // Remove duplicated tokenId
    .filter((tokenId, index, tokenIds) => tokenIds.indexOf(tokenId) === index)
    .filter((tokenId) =>
      restructuredPools.every((pool) => pool.lpToken !== tokenId),
    )
    .map((tokenId) => findOrThrow(tokens, ({ id }) => id === tokenId));
  const nonLpTokenOptions = nonLpTokenSpecs.map((spec) => ({
    tokenId: spec.id,
    ecosystemId: spec.nativeEcosystem,
  }));

  const swimUsdSpec = findOrThrow(tokens, (token) =>
    isSwimUsdTokenId(token.id),
  );
  const swimUsdOptions = [...swimUsdSpec.detailsByEcosystem.keys()].map(
    (ecosystemId) => ({
      tokenId: swimUsdSpec.id,
      ecosystemId,
    }),
  );
  return [...nonLpTokenOptions, ...swimUsdOptions];
};
