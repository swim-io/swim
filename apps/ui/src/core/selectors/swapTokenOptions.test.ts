import {
  DEVNET_POOLS,
  DEVNET_POOLS_FOR_RESTRUCTURE,
  DEVNET_SWIMUSD,
  DEVNET_TOKENS,
  DEVNET_TOKENS_FOR_RESTRUCTURE,
} from "../../config";
import { mockOf } from "../../testUtils";
import { useEnvironment } from "../store";

import { selectConfig } from "./environment";
import { selectSwapTokenOptions } from "./swapTokenOptions";

jest.mock("./environment", () => ({
  ...jest.requireActual("./environment"),
  selectConfig: jest.fn(),
}));

const selectConfigMock = mockOf(selectConfig);

const DISABLED_POOLS = new Set([
  "devnet-meta-acala-ausd",
  "devnet-karura-ausd",
  "devnet-aurora-usn",
]);
const MOCK_POOLS = [...DEVNET_POOLS, ...DEVNET_POOLS_FOR_RESTRUCTURE].filter(
  (pool) => !DISABLED_POOLS.has(pool.id),
);

describe("swapTokenOptions", () => {
  describe("selectSwapTokenOptions", () => {
    beforeEach(() => {
      selectConfigMock.mockReturnValue({
        pools: MOCK_POOLS,
        tokens: [
          ...DEVNET_TOKENS,
          DEVNET_SWIMUSD,
          ...DEVNET_TOKENS_FOR_RESTRUCTURE,
        ],
      });
    });

    it("should return swap options", () => {
      const options = selectSwapTokenOptions(useEnvironment.getState());
      expect(options).toMatchSnapshot();
    });
  });
});
