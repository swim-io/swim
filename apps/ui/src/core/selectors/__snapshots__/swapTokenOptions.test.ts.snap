// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swapTokenOptions selectSwapTokenOptions should return swap options 1`] = `
Array [
  Object {
    "ecosystemId": "solana",
    "tokenId": "devnet-solana-usdc",
  },
  Object {
    "ecosystemId": "solana",
    "tokenId": "devnet-solana-usdt",
  },
  Object {
    "ecosystemId": "ethereum",
    "tokenId": "devnet-ethereum-usdc",
  },
  Object {
    "ecosystemId": "ethereum",
    "tokenId": "devnet-ethereum-usdt",
  },
  Object {
    "ecosystemId": "bnb",
    "tokenId": "devnet-bnb-busd",
  },
  Object {
    "ecosystemId": "bnb",
    "tokenId": "devnet-bnb-usdt",
  },
  Object {
    "ecosystemId": "avalanche",
    "tokenId": "devnet-avalanche-usdc",
  },
  Object {
    "ecosystemId": "avalanche",
    "tokenId": "devnet-avalanche-usdt",
  },
  Object {
    "ecosystemId": "polygon",
    "tokenId": "devnet-polygon-usdc",
  },
  Object {
    "ecosystemId": "polygon",
    "tokenId": "devnet-polygon-usdt",
  },
  Object {
    "ecosystemId": "aurora",
    "tokenId": "devnet-aurora-usdc",
  },
  Object {
    "ecosystemId": "aurora",
    "tokenId": "devnet-aurora-usdt",
  },
  Object {
    "ecosystemId": "fantom",
    "tokenId": "devnet-fantom-usdc",
  },
  Object {
    "ecosystemId": "karura",
    "tokenId": "devnet-karura-usdt",
  },
  Object {
    "ecosystemId": "solana",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "acala",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "aurora",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "avalanche",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "bnb",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "ethereum",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "fantom",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "karura",
    "tokenId": "devnet-swimusd",
  },
  Object {
    "ecosystemId": "polygon",
    "tokenId": "devnet-swimusd",
  },
]
`;
